guja-core/src/main/java/com/wadpam/guja/cache/CacheMethodInterceptor.java

```diff
@@ -0,0 +1,188 @@
+
+    if (PUT_METHOD_NAME.equals(method.getName()) && method.isAnnotationPresent(Crud.class)) {
+      LOGGER.trace("   put");
+      final Object id = invocation.proceed();
+      args[1] = id;
+      final Object entity = args[2];
+      args[2] = null;
+      checkNotNull(id);
+      checkNotNull(entity);
+      cache.put(new Triple(args), Optional.of(entity));
+      return id;
+    }
+
+    if (DELETE_METHOD_NAME.equals(method.getName()) && method.isAnnotationPresent(Crud.class)) {
+      LOGGER.info("   delete");
+      invocation.proceed();
+      final Object id = args[1];
+      checkNotNull(id);
+      cache.put(triple, Optional.absent());
+      return null;
+    }
+
+    final Optional<?> optionalEntity = cache.get(triple, new Callable<Optional<?>>() {
+      @Override
+      public Optional<?> call() throws Exception {
+        LOGGER.trace("Loading for {}({})", method.getName(), triple);
+        try {
+          final Object entity = invocation.proceed();
+          return null != entity ? Optional.of(entity) : Optional.absent();
+        } catch (Throwable throwable) {
+          LOGGER.error("Failed to populate cache {} {}", clazz.getName(), throwable);
+          throw new ExecutionException("During invocation: ", throwable);
+        }
+      }
+    });
+    return null != optionalEntity && optionalEntity.isPresent() ? optionalEntity.get() : null;
+  }
+
+  private Cache<Triple, Optional<?>> getCacheInstance(Class clazz) {
+    final String className = clazz.getName();
+
+    Cache<Triple, Optional<?>> cache = namespaces.get(className);
+    if (null == cache) {
+      final Cached annotation = (Cached) clazz.getAnnotation(Cached.class);
+      LOGGER.debug("Build new dao cache for {}", className);
+      CacheBuilder<Triple, Optional<?>> cacheBuilder = cacheBuilderProvider.get();
+      if (null != annotation.from() || "".equals(annotation.from())) {
+        cacheBuilder.from(annotation.from());
+      } else {
+        cacheBuilder.from(className);
+      }
+      if (annotation.size() > 0) {
+        cacheBuilder.maximumSize(annotation.size());
+      }
+      if (annotation.expiresAfterSeconds() > 0) {
+        cacheBuilder.expireAfterWrite(annotation.expiresAfterSeconds());
+      }
+      cache = cacheBuilder.build();
+      namespaces.put(className, cache);
+    }
+
+    return cache;
+  }
+
+  private Class getClass(MethodInvocation invocation) {
+    Class clazz = invocation.getThis().getClass();
+    if (clazz.getName().contains("$$EnhancerByGuice$$")) {
+      // Workaround for Guice adding some kind of enhancer object before the actual DAO
+      // Check the superclass
+      clazz = clazz.getSuperclass();
+      if (!clazz.isAnnotationPresent(Cached.class)) {
+        throw new InternalServerErrorRestException("Could not find Cached annotation");
+      }
+    }
+    return clazz;
+  }
+
+  public static class Triple implements Serializable {
```

Comment:
There is a Triplet implementation in the base module. Lets keep this and I will move the implementation from base and merge with this Triplet implementation.